### PR TITLE
docs: Clarify versioning and API stability

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
As shared via Slack, we’re seeing a few concerns from the community on API stability in minor and patch releases. We should clarify those so that users can decide whether they rely on certain API in their code.

### Key points to clarify
- PL does not follow [SemVer](https://semver.org/).
- Using the latest patch release within a minor release is always safe (except for experimental/beta features).
- (... any suggestions are welcome!)

### Example
- pandas: https://pandas.pydata.org/docs/development/policies.html#version-policy
- Flash: https://lightning-flash.readthedocs.io/en/latest/stability.html
- React: https://reactjs.org/docs/faq-versioning.html

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @Felonious-Spellfire @akihironitta

### CODE REVIEW

File: docs/source-pytorch/release_policy.rst [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+maximizing ease of transition from one version to the next. We take backward compatibility and reproducibility very
+seriously.
+
+For API removal, renaming or other forms of backward-incompatible changes, the procedure is:
+
+#. A deprecation process is initiated at a minor version ``X``, producing warning messages at runtime and in the documentation.
+#. Calls to the deprecated API remain unchanged in their function during the deprecation phase.
+#. Two minor versions in the future at version ``X+2`` the breaking change takes effect.
+
+The "X+2" rule is a recommendation and not a strict requirement. Longer deprecation cycles may apply for some cases.
```
Comment (akihironitta) [inline docs/source-pytorch/release_policy.rst:34]: > The "X+2" rule is a recommendation and not a strict requirement. Longer deprecation cycles may apply for some cases.

In what situation, it is likely that we don't follow this rule?

Also, I see there might be longer deprecation cycles such as for [these](https://github.com/Lightning-AI/lightning/blob/273a9ed8c14650c59b60c546fe53d93a4183d439/tests/tests_pytorch/deprecated_api/test_remove_2-0.py) but do we have **shorter** deprecations, too? If yes, we need to document it as well.

**Review by akihironitta — COMMENTED** [review]
(kind of off-topic, but )Would it be make sense to document the window size of old PyTorch support here?

File: docs/source-pytorch/release_policy.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source-pytorch/release_policy.rst:34]: Speaking from what I experienced of the past: Default is X+2 and on a case-by-case basis, deviations sometimes get discussed. 

Example for a shorter deprecation:
- Function is broken/unusable beyond repair, and gets deprecated or changed. Shorter period because the feature is not usable in the current version. (DDP2 was a concrete example)

Example for a longer deprecation:
- An API that is so core to the Lightning experience that it is hard and time-consuming to make the switch for everyone involved, user and developer (us): Trainer(gpus=x) -> Trainer(devices=x), Trainer.fit() -> Trainer.train() etc. The user gets into contact with these APIs in almost all use cases.

**Review by awaelchli — APPROVED** [review]

File: docs/source-pytorch/governance.rst [hunk 2]
```diff
@@ -57,15 +57,6 @@ Alumni
 - Peter Yu (`yukw777 <https://github.com/yukw777>`_)
 - Roger Shieh (`s-rog <https://github.com/s-rog>`_)
 
-
-Releases
```
Comment (akihironitta) [inline docs/source-pytorch/governance.rst:61]: "Releases" and "API Evolution" sections are moved to a new standalone versioning policy page with improvements.

File: docs/source-pytorch/versioning.rst [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+#. The deprecation warning lasts for two minor versions before the breaking change takes effect at ``X+2`` minor version.
+
+If your code raises no deprecation warning with the latest minor release, it will be compatible with the next minor release.
+
+The ``X+2`` rule is a recommendation and not a strict requirement. Shorter or longer deprecation cycles may apply to some cases. In the past, DDP2 was removed without a deprecation process because the feature was broken and unusable beyond fixing as discussed in `#12584 <https://github.com/Lightning-AI/lightning/issues/12584>`_. Also, `#10410 <https://github.com/Lightning-AI/lightning/issues/10410>`_ is an example where a longer deprecation applied. We deprecated the accelerator arguments, such as ``Trainer(gpus=...)``, in 1.7, however, because the APIs were so core that they would impact almost all use cases, we decided not to introduce the breaking change until 2.0.
+
+Python Support
+**************
+
+PyTorch Lightning follows `NEP 29 <https://numpy.org/neps/nep-0029-deprecation_policy.html>`_ which PyTorch also follows (`#74203 <https://github.com/pytorch/pytorch/issues/74203>`_).
```
Comment (akihironitta) [inline docs/source-pytorch/versioning.rst:61]: Please mind that this hasn't been confirmed at all by the team yet. Is it reasonable to officially mention which Python versions we support?

File: docs/source-pytorch/versioning.rst [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+PyTorch Support
+***************
+
+PyTorch Lightning supports the latest four minor versions of PyTorch. For example, PyTorch Lightning 1.7 supports PyTorch 1.9, 1.10, 1.11 and 1.12.
```
Comment (akihironitta) [inline docs/source-pytorch/versioning.rst:66]: I've seen our PyTorch version support window mentioned in the past issues/PRs by some core members. Can we officially state the rule here?

File: docs/source-pytorch/versioning.rst [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+----------------
+
+Experimental APIs are labelled as experimental or beta in the documentation and/or in the release note and are considered unstable and should not be used in production.
+
+For experimental features, any of the following may be true.
+
+- The feature has unstable dependencies.
+- The API may change without notice in future versions.
+- The performance of the feature has not been verified.
+- The docs for this feature are under active development.
```
Comment (akihironitta) [inline docs/source-pytorch/versioning.rst:33]: These bullet points are basically copied from Flash's stability note written by @ethanwharris.

File: docs/source-pytorch/versioning.rst (hunk 4 above)
Comment (carmocca) [reply docs/source-pytorch/versioning.rst:66]: IMO yes, but I'm not sure if 1.8 is going to break this pattern. Can you clarify @Borda?

**Review by carmocca — APPROVED** [review]

File: docs/source-pytorch/versioning.rst (hunk 4 above)
Comment (Borda) [reply docs/source-pytorch/versioning.rst:66]: I think we keep this window

**Review by Borda — APPROVED** [review]

**Review by otaj — APPROVED** [review]
